samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -315,4 +315,48 @@ public void onBarrierError(String version, Throwable t) {
       stop();
     }
   }
+
+  /// listener to handle session expiration
+  public class ZkSessionStateChangedListener implements IZkStateListener {
+
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state)
+        throws Exception {
+      if (state == Watcher.Event.KeeperState.Expired) {
+        // if the session has expired it means that all the registration's ephemeral nodes are gone.
+        LOG.warn("Got session expired event for processor=" + processorId);
+      }
+    }
+
+    @Override
+    public void handleNewSession()
+        throws Exception {
+      LOG.info("Got new session created event for processor=" + processorId);
+
+      // increase generation of the ZK connection. All the callbacks from the previous generation will be ignored.
+      zkUtils.incGeneration();
+
+      if (coordinatorListener != null) {
+        coordinatorListener.onJobModelExpired();
```

Comment:
onJobModelExpired is a blocking call. Should that be handled directly in the event thread? Esp since shutting down the container is taking longer for certain jobs?